crates/bevy_reflect/bevy_reflect_derive/src/type_uuid.rs

```diff
@@ -50,24 +51,70 @@ pub(crate) fn type_uuid_derive(input: proc_macro::TokenStream) -> proc_macro::To
 
     let uuid =
         uuid.expect("No `#[uuid = \"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx\"` attribute found.");
+    gen_impl_type_uuid(TypeUuidDef {
+        type_name: name,
+        generics: ast.generics,
+        uuid,
+    })
+}
+
+/// Generates an implementation of [`TypeUuid`][crate::TypeUuid]. If there any generics, the `TYPE_UUID` will be a composite of the generic types' `TYPE_UUID`.
+pub(crate) fn gen_impl_type_uuid(def: TypeUuidDef) -> proc_macro::TokenStream {
+    let uuid = def.uuid;
+    let mut generics = def.generics;
+    let type_name = def.type_name;
+
+    let bevy_reflect_path: Path = BevyManifest::default().get_path("bevy_reflect");
+
+    generics.type_params_mut().for_each(|param| {
+        param
+            .bounds
+            .push(syn::parse_quote!(#bevy_reflect_path::TypeUuid));
+    });
+
     let bytes = uuid
         .as_bytes()
         .iter()
         .map(|byte| format!("{byte:#X}"))
         .map(|byte_str| syn::parse_str::<LitInt>(&byte_str).unwrap());
 
+    let (impl_generics, type_generics, where_clause) = generics.split_for_impl();
     let base = quote! { #bevy_reflect_path::Uuid::from_bytes([#( #bytes ),*]) };
-    let type_uuid = ast.generics.type_params().fold(base, |acc, param| {
+
+    let type_uuid = generics.type_params().fold(base, |acc, param| {
         let ident = &param.ident;
         quote! {
             #bevy_reflect_path::__macro_exports::generate_composite_uuid(#acc, <#ident as #bevy_reflect_path::TypeUuid>::TYPE_UUID)
         }
     });
 
     let gen = quote! {
-        impl #impl_generics #bevy_reflect_path::TypeUuid for #name #type_generics #where_clause {
+        impl #impl_generics #bevy_reflect_path::TypeUuid for #type_name #type_generics #where_clause {
             const TYPE_UUID: #bevy_reflect_path::Uuid = #type_uuid;
         }
     };
     gen.into()
 }
+
+/// A struct containing the data required to generate an implementation of [`TypeUuid`][crate::TypeUuid].
+pub(crate) struct TypeUuidDef {
+    pub type_name: Ident,
+    pub generics: Generics,
+    pub uuid: Uuid,
+}
+/// Parses the data to be passed into [`impl_type_uuid!`][crate::impl_type_uuid!]. This should be in the format of `[Type]<[Generic Params]>, [Uuid as a literal u128]`
```

Comment:
Nit: This should probably go on the actual `TypeUuidDef` type definition rather than on the `Parse` impl.

This should probably also be included as a doc comment on the exported `impl_type_uuid!` macro itself.